google-ads-examples/src/main/java/com/google/ads/googleads/examples/errorhandling/HandleKeywordPolicyViolations.java

```diff
@@ -0,0 +1,214 @@
+    } catch (GoogleAdsException gae) {
+      // GoogleAdsException is the base class for most exceptions thrown by an API request.
+      // Instances of this exception have a message and a GoogleAdsFailure that contains a
+      // collection of GoogleAdsErrors that indicate the underlying causes of the
+      // GoogleAdsException.
+      System.err.printf(
+          "Request ID %s failed due to GoogleAdsException. Underlying errors:%n",
+          gae.getRequestId());
+      int i = 0;
+      for (GoogleAdsError googleAdsError : gae.getGoogleAdsFailure().getErrorsList()) {
+        System.err.printf("  Error %d: %s%n", i++, googleAdsError);
+      }
+    }
+  }
+
+  /**
+   * Runs the example.
+   *
+   * @param googleAdsClient the Google Ads API client.
+   * @param customerId the client customer ID.
+   * @param adGroupId the ID of the ad group to add a keyword to.
+   * @param keywordText the keyword text to add.
+   * @throws GoogleAdsException if an API request failed with one or more service errors.
+   */
+  private void runExample(
+      GoogleAdsClient googleAdsClient, long customerId, Long adGroupId, String keywordText) {
+    try (AdGroupCriterionServiceClient adGroupCriterionServiceClient =
+        googleAdsClient.getLatestVersion().createAdGroupCriterionServiceClient()) {
+      // Configures the keyword text and match type settings.
+      KeywordInfo keywordInfo =
+          KeywordInfo.newBuilder()
+              .setText(StringValue.of(keywordText))
+              .setMatchType(KeywordMatchType.EXACT)
+              .build();
+
+      // Constructs an ad group criterion using the keyword text info above.
+      AdGroupCriterion adGroupCriterion =
+          AdGroupCriterion.newBuilder()
+              .setAdGroup(StringValue.of(ResourceNames.adGroup(customerId, adGroupId)))
+              .setStatus(AdGroupCriterionStatus.ENABLED)
+              .setKeyword(keywordInfo)
+              .build();
+
+      // Constructs an operation to create the ad group criterion.
+      AdGroupCriterionOperation operation =
+          AdGroupCriterionOperation.newBuilder().setCreate(adGroupCriterion).build();
+
+      // Tries sending a mutate request to add the keyword.
+      List<PolicyViolationKey> exemptibleKeys = new ArrayList<>();
+      int errorCount = 0;
+      try {
+        MutateAdGroupCriteriaResponse response =
+            adGroupCriterionServiceClient.mutateAdGroupCriteria(
+                String.valueOf(customerId), Arrays.asList(operation));
+      } catch (GoogleAdsException e) {
+        exemptibleKeys = extractExemptiblePolicyViolationKeys(e);
+        errorCount = e.getGoogleAdsFailure().getErrorsCount();
+      }
+
+      // Tries sending exemption requests for creating the keyword. However, if your keyword
+      // contains many policy violations, but not all of them are exemptible, the request will not
+      // be sent.
+      if (exemptibleKeys.size() == errorCount) {
```

Comment:
Shouldn't we check if the number of errors > 0?

Reply:
Good point. I thought this would get a bit confusing with if/else-if/else, so instead I added an early `return` above with some comments to cover the case where the first request succeeds.